### ISSUE TITLE
ARM Templates for Event Subscriptions

### ISSUE THREAD

**Issue by ciappara** [Azure/azure-resource-manager-schemas#563]
Hi everyone,

I'm having an issue with ARM Templates. I'm adding an Event Subscription in IoT Hub, from the Events tab, to receive the Device Connected and Disconnected events. That works fine, but I want to create the Event Subscriptions using an ARM Template.

![image](https://user-images.githubusercontent.com/11292606/51525478-ba0a2000-1e30-11e9-9f36-04ef289eb14f.png)

In the Resource Group, the Automation Script is not updated after that the Event Subscription is created, so I cannot get the ARM template code that creates the event subscriptions.

I then copied the configuration from the Advanced Editor when creating a new Event Subscription...:

![image](https://user-images.githubusercontent.com/11292606/51530785-07d85580-1e3c-11e9-836d-7af2c563c4ae.png)

...and added it as part of the ARM template:

```
      "type": "Microsoft.EventGrid/eventSubscriptions",
      "location": "westeurope",
      "apiVersion": "2018-09-15-preview",
      "name": "ConnectionStateEventSubscription",
      "properties": {
        "topic": "/subscriptions/xxxxx-xxxx-xxx-xxxxx/resourceGroups/resource-group/providers/Microsoft.Devices/IotHubs/iot-hub",
        "destination": {
          "endpointType": "WebHook",
          "properties": {
            "endpointUrl": "https://url-to.endpoint/webhook"
          }
        },
        "filter": {
          "includedEventTypes": [
            "Microsoft.Devices.DeviceConnected",
            "Microsoft.Devices.DeviceDisconnected"
          ],
          "advancedFilters": []
        },
        "labels": [],
        "eventDeliverySchema": "EventGridSchema"
```

But when I run the ARM template from Visual Studio (through Powershell) I'm receiving following error:

![image](https://user-images.githubusercontent.com/11292606/51538900-bbe4db00-1e52-11e9-94ab-b428c10070cb.png)

How should the configuration in the ARM templates be set to create Event Subscriptions in an IoT Hub?

Thank you in advance,
Clive

**Comment by jwilliams-git** [comment]
I don't see a _topic_ property here - https://docs.microsoft.com/en-us/azure/templates/microsoft.eventgrid/2018-09-15-preview/eventsubscriptions .

Try this:

```
{
      "type": "Microsoft.EventGrid/topics/providers/eventSubscriptions",
      "name": "iot-hub/Microsoft.EventGrid/ConnectionStateEventSubscription",
      "location": "westeurope",
      "apiVersion": "2018-09-15-preview",
      "properties": {
        "destination": {
          "endpointType": "WebHook",
          "properties": {
            "endpointUrl": "https://url-to.endpoint/webhook"
          }
        },
        "filter": {
          "includedEventTypes": [
            "Microsoft.Devices.DeviceConnected",
            "Microsoft.Devices.DeviceDisconnected"
          ]
        }
      }
}
```

**Comment by ciappara** [comment]
Hi JWilliams,

Thank you for your quick reply. I updated the arm template and ran it, but received the following error (iot-clivearm is the IoT Hub name):

```
15:11:09 - Template deployment returned the following errors:
15:11:09 - 15:11:08 - Resource Microsoft.EventGrid/topics/providers/eventSubscriptions 'iot-clivearm/Microsoft.EventGrid/ConnectionStateEventSubscription' failed with message '{
15:11:09 -   "error": {
15:11:09 -     "code": "ResourceNotFound",
15:11:09 -     "message": "The Resource 'Microsoft.EventGrid/topics/iot-clivearm' under resource group 'iot-clivearm' was not found."
15:11:09 -   }
15:11:09 - }'
```

Do you know what the problem might be? Thank you in advance.

**Comment by jwilliams-git** [comment]
Sorry, I misread your original post. I tried a couple of other ideas that I had including a rest-api request just as a poc. None of my ideas worked; I can't seem to get past the error _Invalid event subscription request: endpoint._. When I get some time, I'll keep digging.

**Comment by ciappara** [comment]
I don't think I received that error. Not sure if this relates to it, but for the webhook endpoint, I had to return a ValidationResponse from the POST method, as Event Grid needs to receive the validation code sent to pass its validation process (Logic Apps take care of validation responses on their own).

Thank you so much for your time!

**Comment by jwilliams-git** [comment]
Yes, you're exactly right! If you already have an endpoint configured to complete the _validationCode/validationResponse_ handshake with event grid then this might be what you're looking for (in its simplest form):

```
{
  "type": "Microsoft.EventGrid/eventSubscriptions",
  "name": "ConnectionStateEventSubscription",
  "location": "westeurope",
  "apiVersion": "2018-01-01",
  "properties": {
    "destination": {
      "endpointType": "WebHook",
      "properties": {
        "endpointUrl": "https://url-to.endpoint/webhook"
      }
    },
    "filter": {
      "includedEventTypes": [
        "Microsoft.Devices.DeviceConnected",
        "Microsoft.Devices.DeviceDisconnected"
      ]
    }
  }
}
```

**Comment by syamkp1986** [comment]
Hi Clive Ciappara ,

Could you find any solution for this? I am also struggling with the same issue.

**Comment by ahamad-MS** [comment]
Hello @ciappara 

Thanks for reporting this issue. I can see a couple of issue in the template you are using:
1. type should should be specified and have value: "Microsoft.Devices/IotHubs/providers/eventSubscriptions",

2. name should take the format: "[IotHubName]/Microsoft.EventGrid/[EventSubscriptionName]" and not the event subscription name only


Please use the following template to create an Event Subscription to an IOT Hub (derived from this template for creating event subscription to storage account posted here: https://github.com/Azure/azure-quickstart-templates/blob/master/101-event-grid-subscription-and-storage/azuredeploy.json). This template uses event hub as destination but you can replace that with webhook too.

{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "resources": 
    [
        {
            "type": "Microsoft.Devices/IotHubs/providers/eventSubscriptions",
            "name": "[IotHubName]/Microsoft.EventGrid/[EventSubscriptionName]",
            "apiVersion": "2018-01-01",
            "properties": 
            {
		"topic": "/subscriptions/[SubscriptionID]/resourcegroups/[ResourceGroup]/providers/Microsoft.Devices/IotHubs/[IotHubName]",
                "destination":
                {
                    "endpointType": "EventHub",
                    "properties": 
                    {
		        "resourceId": "/subscriptions/[SubscriptionId]/resourceGroups/[ResourceGroup]/providers/Microsoft.EventHub/namespaces/[EventHubNameSpace]/eventhubs/[EventHubName]"
                    }
                },
		"filter": 
                {
		    "subjectBeginsWith": "xyz",
	            "subjectEndsWith": "vwx",
		    "isSubjectCaseSensitive": true,
		    "includedEventTypes": 
                    [
		        "Microsoft.Devices.DeviceCreated",
			"Microsoft.Devices.DeviceDeleted"
		    ],
		    "advancedFilters": []
		},
		"labels": [],
		"eventDeliverySchema": "EventGridSchema"            
           }
        }
    ]
}

Please let us know if you still hit any more issues. thanks
